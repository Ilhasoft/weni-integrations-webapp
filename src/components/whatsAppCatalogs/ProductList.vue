<template>
  <div class="whatsapp-product-list">
    <div class="whatsapp-product-list__header">
      <div class="whatsapp-product-list__header__text">
        <div class="whatsapp-product-list__header__icon">
          <img src="@/assets/svgs/storefront.svg" />
        </div>
        <div class="whatsapp-product-list__header__title">
          <span class="u font primary title-sm color-neutral-darkest">
            {{ catalogName }}
          </span>
          <span class="u font secondary body-gt color-neutral-darkest">
            {{ $t('WhatsApp.product.list.description') }}
          </span>
        </div>
      </div>
    </div>

    <unnnic-table :items="listItems" class="whatsapp-product-list__table">
      <template v-slot:header>
        <unnnic-table-row :headers="headers" />
      </template>

      <template v-slot:item="{ item }">
        <unnnic-table-row :headers="headers">
          <template v-slot:title>
            <div class="whatsapp-product-list__table__title">
              <img :src="item.image_link" class="whatsapp-product-list__table__title__image" />
              <span :data-title="item.title">{{ item.title }}</span>
            </div>
          </template>

          <template v-slot:available>
            <div :title="item.available" class="break-text whatsapp-product-list__table__status">
              <span :class="`whatsapp-product-list__table__status__indicator ${item.available}`" />
              <span>{{ item.available ? 'Disponível' : 'Indisponível' }}</span>
            </div>
          </template>

          <template v-slot:price>
            <div :title="item.price" class="break-text whatsapp-product-list__table__price">
              <span>{{ item.price }}</span>
              <span
                v-if="item.base_price !== item.price"
                class="whatsapp-product-list__table__price__discount"
              >
                {{ item.base_price }}
              </span>
            </div>
          </template>

          <template v-slot:facebook_product_id>
            <div :title="item.facebook_product_id" class="break-text">
              {{ item.facebook_product_id }}
            </div>
          </template>
        </unnnic-table-row>
      </template>
    </unnnic-table>
    <div class="whatsapp-product-list__pagination">
      <span>{{ currentPageStart }} - {{ currentPageCount }} de {{ totalCount }}</span>
      <unnnic-pagination
        :modelValue="page"
        @update:modelValue="onPageChange"
        :max="pageCount"
        :show="5"
      />
    </div>
  </div>
</template>

<script>
  import { mapActions, mapState } from 'pinia';
  import { whatsapp_cloud } from '@/stores/modules/appType/channels/whatsapp_cloud.store';

  export default {
    name: 'ProductList',
    props: {
      catalogName: {
        type: String,
        required: true,
        default: '',
      },
    },
    data() {
      return {
        firstLoad: true,
        page: 1,
        pageSize: 15,
        headers: [
          {
            id: 'title',
            text: this.$t('WhatsApp.product.table.headers.name'),
            flex: 2,
          },
          {
            id: 'available',
            text: this.$t('WhatsApp.product.table.headers.availability'),
            flex: 1,
          },
          {
            id: 'price',
            text: this.$t('WhatsApp.product.table.headers.price'),
            flex: 1,
          },
          {
            id: 'facebook_product_id',
            text: this.$t('WhatsApp.product.table.headers.facebook_id'),
            flex: 1,
          },
        ],
      };
    },
    async mounted() {
      await this.fetchProducts(this.page);
    },
    computed: {
      ...mapState(whatsapp_cloud, [
        'catalogProducts',
        'loadingCatalogProducts',
        'errorCatalogProducts',
      ]),
      listItems() {
        return this.catalogProducts?.results || [];
      },
      totalCount() {
        return this.catalogProducts?.count || this.pageSize;
      },
      pageCount() {
        if (this.catalogProducts?.count) {
          return Math.ceil(this.catalogProducts.count / this.pageSize);
        } else {
          return 1;
        }
      },
      currentPageStart() {
        return (this.page - 1) * this.pageSize || 1;
      },
      currentPageCount() {
        const value = this.pageSize * this.page;
        return value > this.catalogProducts?.count ? this.catalogProducts?.count || 0 : value;
      },
    },
    methods: {
      ...mapActions(whatsapp_cloud, ['getCatalogProducts']),
      async fetchProducts(page) {
        const { appUuid, catalogUuid } = this.$route.params;
        const params = {
          page: page,
          page_size: this.pageSize,
        };

        await this.getCatalogProducts({ appUuid, catalogUuid, params });
      },
      onPageChange(value) {
        this.page = value;
        this.currentPage = value;
      },
    },
    watch: {
      page: {
        immediate: true,
        handler(page) {
          this.fetchProducts(page);
        },
      },
    },
  };
</script>

<style lang="scss" scoped>
  .whatsapp-product-list {
    display: flex;
    flex-direction: column;
    flex: 1;
    overflow: hidden;
    gap: $unnnic-spacing-md;

    &__header {
      display: flex;
      align-items: center;
      justify-content: space-between;

      &__icon {
        display: flex;
        background-color: #eefffc;
        border-radius: $unnnic-border-radius-sm;
        justify-content: center;
        margin-right: $unnnic-spacing-inline-sm;

        img {
          height: $unnnic-icon-size-xl;
          width: $unnnic-icon-size-xl;
          padding: $unnnic-inset-nano;
        }
      }
      &__text {
        display: flex;
      }
      &__title {
        display: flex;
        flex-direction: column;
        gap: $unnnic-spacing-nano;
      }
    }

    &__search {
      width: 33%;
      min-width: 250px;
      max-width: 450px;
    }

    &__cards {
      display: flex;
      flex-direction: column;
      flex: 1;
      gap: $unnnic-spacing-md;
      overflow: auto;
      padding-right: $unnnic-spacing-xs;
    }

    &__pagination {
      display: flex;
      justify-content: space-between;
      align-items: center;

      span {
        font-size: $unnnic-font-size-body-gt;
        line-height: $unnnic-line-height-md + $unnnic-font-size-body-gt;
        color: $unnnic-color-neutral-dark;
      }
    }

    ::v-deep .unnnic-modal .container .content {
      padding-right: 0px;
    }

    ::v-deep .unnnic-modal.type-alert .title {
      padding-bottom: $unnnic-spacing-xs;
    }

    ::v-deep .unnnic-modal.type-alert .container .content.with-validation .description {
      margin-bottom: $unnnic-spacing-sm;
    }

    &__table {
      display: flex;
      height: 100%;
      overflow-x: hidden;
      overflow-y: auto;

      ::v-deep .scroll {
        overflow-x: hidden;
        overflow-y: auto;

        padding-right: unset;
      }

      &__title {
        display: flex;
        align-items: center;

        // add ellipsis if it breaks text
        span {
          overflow: hidden;
          text-overflow: ellipsis;
          white-space: nowrap;
        }

        &__image {
          height: $unnnic-avatar-size-sm;
          width: $unnnic-avatar-size-sm;
          border-radius: $unnnic-border-radius-sm;
          margin-right: $unnnic-spacing-inline-sm;
          object-fit: cover;
          aspect-ratio: 1 / 1;
        }
      }

      &__status {
        display: flex;
        align-items: center;
        gap: $unnnic-spacing-nano;

        &__indicator {
          display: inline-block;
          height: $unnnic-spacing-xs;
          width: $unnnic-spacing-xs;
          margin: 0 6px;
          border-radius: $unnnic-border-radius-pill;
          &.false {
            background-color: $unnnic-color-aux-red-500;
          }

          &.true {
            background-color: $unnnic-color-aux-green-500;
          }
        }
      }

      &__price {
        display: flex;
        gap: $unnnic-spacing-nano;
        &__discount {
          text-decoration: line-through;
          color: $unnnic-color-neutral-cleanest;
        }
      }
    }

    &__pagination {
      display: flex;
      justify-content: space-between;
      align-items: center;
      margin-top: $unnnic-spacing-stack-md;

      span {
        font-size: $unnnic-font-size-body-gt;
        line-height: $unnnic-line-height-md + $unnnic-font-size-body-gt;
        color: $unnnic-color-neutral-dark;
      }
    }
  }
</style>
