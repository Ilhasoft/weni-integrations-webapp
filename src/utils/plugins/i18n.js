import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

import en from '../../locales/en.json';
import pt_br from '../../locales/pt_br.json';
import es_es from '../../locales/es_es.json';

const languages = {
  'en-us': en,
  'pt-br': pt_br,
  es: es_es,
};

const messages = Object.assign(languages);

const i18n = new VueI18n({
  locale: 'en-us',
  fallbackLocale: 'en-us',
  messages,
});

export default i18n;
