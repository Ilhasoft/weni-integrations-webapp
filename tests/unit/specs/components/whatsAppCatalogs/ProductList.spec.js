import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import ProductList from '@/components/whatsAppCatalogs/ProductList.vue';
import '@weni/unnnic-system';
import i18n from '@/utils/plugins/i18n';

const localVue = createLocalVue();
localVue.use(Vuex);

const mountComponent = async ({
  catalogName = 'catalog name',
  catalogProducts = {
    results: [
      {
        image_link: 'image-link',
        title: 'title',
        available: true,
        price: 'R$ 10,00',
        base_price: 'R$ 12,00',
        facebook_product_id: 'fb-id-123',
      },
    ],
    count: 1,
  },
  loadingCatalogProducts = false,
  errorCatalogProducts = null,
} = {}) => {
  const state = {
    catalogProducts,
    loadingCatalogProducts,
    errorCatalogProducts,
  };

  const actions = {
    getCatalogProducts: jest.fn(),
  };

  const store = new Vuex.Store({
    modules: {
      WhatsAppCloud: {
        namespaced: true,
        actions,
        state,
      },
    },
  });

  const wrapper = mount(ProductList, {
    localVue,
    store,
    i18n,
    propsData: {
      catalogName,
    },
    mocks: {
      $route: {
        params: {
          appUuid: 'app-uuid-123',
          catalogUuid: 'catalog-uuid-123',
        },
      },
    },
  });

  return { wrapper, actions, state };
};

describe('components/whatsAppCatalog/ProductList.vue', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should be rendered properly', async () => {
    const { wrapper } = await mountComponent();
    expect(wrapper).toMatchSnapshot();
  });

  describe('computed', () => {
    describe('listItems', () => {
      it('should return an empty array if there are no catalogs', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: null });
        expect(wrapper.vm.listItems).toEqual([]);
      });

      it('should return an array of catalog products', async () => {
        const { wrapper } = await mountComponent();
        expect(wrapper.vm.listItems).toEqual(wrapper.vm.catalogProducts.results);
      });
    });

    describe('totalCount', () => {
      it('should return pageSize if there are no catalogs provided', async () => {
        const { wrapper } = await mountComponent({
          catalogProducts: { results: [], count: 0 },
        });
        expect(wrapper.vm.totalCount).toBe(wrapper.vm.pageSize);
      });
    });

    describe('pageCount', () => {
      it('should return 1 if there are no catalogs', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: { results: [], count: 0 } });
        expect(wrapper.vm.pageCount).toBe(1);
      });

      it('should return correct number of pages based on limit and count', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: { results: [], count: 30 } });
        expect(wrapper.vm.pageCount).toBe(2);
      });
    });

    describe('currentPageStart', () => {
      it('should return 1 if its the first page', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: { results: [], count: 0 } });
        expect(wrapper.vm.currentPageStart).toBe(1);
      });

      it('should return page limit start based on the current page', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: { results: [], count: 30 } });
        await wrapper.setData({ page: 2 });
        expect(wrapper.vm.currentPageStart).toBe(wrapper.vm.pageSize);
      });
    });

    describe('currentPageCount', () => {
      it('should return currentPageCount as 0 if there are not enough templates', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: { results: [], count: 0 } });
        expect(wrapper.vm.currentPageCount).toBe(0);
      });

      it('should return current number of templates if there are more than one and less than page limit and page equals one', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: { results: [], count: 2 } });
        expect(wrapper.vm.currentPageCount).toBe(2);
      });

      it('should return the pageSize as count if there are enough templates to show', async () => {
        const { wrapper } = await mountComponent({ catalogProducts: { results: [], count: 23 } });
        expect(wrapper.vm.currentPageCount).toBe(wrapper.vm.pageSize);
      });
    });
  });
});
