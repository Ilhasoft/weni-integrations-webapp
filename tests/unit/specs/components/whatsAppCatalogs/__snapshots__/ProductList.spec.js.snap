// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/whatsAppCatalog/ProductList.vue should be rendered properly 1`] = `
<div class="whatsapp-product-list">
  <div class="whatsapp-product-list__header">
    <div class="whatsapp-product-list__header__text">
      <div class="whatsapp-product-list__header__icon"><img src="@/assets/svgs/storefront.svg"></div>
      <div class="whatsapp-product-list__header__title"><span class="u font primary title-sm color-neutral-darkest">
          catalog name
        </span> <span class="u font secondary body-gt color-neutral-darkest">
          Here you manage your inventory information. Add all the items you want to advertise or sell
        </span></div>
    </div>
  </div>
  <div data-v-cb43f82a="" class="unnnic-table whatsapp-product-list__table">
    <div data-v-cb43f82a="" class="header">
      <div data-v-35dce591="" class="table-row" data-v-cb43f82a="">
        <div data-v-35dce591="" class="col" style="flex: 2; min-width: 0;">
          <div data-v-35dce591="" class="break-text">Name</div>
        </div>
        <div data-v-35dce591="" class="divider"></div>
        <div data-v-35dce591="" class="col" style="flex: 1; min-width: 0;">
          <div data-v-35dce591="" class="break-text">Availability</div>
        </div>
        <div data-v-35dce591="" class="divider"></div>
        <div data-v-35dce591="" class="col" style="flex: 1; min-width: 0;">
          <div data-v-35dce591="" class="break-text">Price</div>
        </div>
        <div data-v-35dce591="" class="divider"></div>
        <div data-v-35dce591="" class="col" style="flex: 1; min-width: 0;">
          <div data-v-35dce591="" class="break-text">Facebook ID</div>
        </div>
        <div data-v-35dce591="" class="divider" style="display: none;"></div>
      </div>
    </div>
    <div data-v-cb43f82a="">
      <div data-v-cb43f82a="" class="item-spacer"></div>
    </div>
    <div data-v-cb43f82a="" class="scroll">
      <div data-v-cb43f82a="" class="item">
        <div data-v-35dce591="" class="table-row" data-v-cb43f82a="">
          <div data-v-35dce591="" class="col" style="flex: 2; min-width: 0;">
            <div data-v-35dce591="" class="whatsapp-product-list__table__title"><img data-v-35dce591="" src="image-link" class="whatsapp-product-list__table__title__image"> <span data-v-35dce591="" data-title="title">title</span></div>
          </div>
          <div data-v-35dce591="" class="divider"></div>
          <div data-v-35dce591="" class="col" style="flex: 1; min-width: 0;">
            <div data-v-35dce591="" title="true" class="break-text whatsapp-product-list__table__status"><span data-v-35dce591="" class="whatsapp-product-list__table__status__indicator true"></span> <span data-v-35dce591="">Dispon√≠vel</span></div>
          </div>
          <div data-v-35dce591="" class="divider"></div>
          <div data-v-35dce591="" class="col" style="flex: 1; min-width: 0;">
            <div data-v-35dce591="" title="R$ 10,00" class="break-text whatsapp-product-list__table__price"><span data-v-35dce591="">R$ 10,00</span> <span data-v-35dce591="" class="whatsapp-product-list__table__price__discount">
              R$ 12,00
            </span></div>
          </div>
          <div data-v-35dce591="" class="divider"></div>
          <div data-v-35dce591="" class="col" style="flex: 1; min-width: 0;">
            <div data-v-35dce591="" title="fb-id-123" class="break-text">
              fb-id-123
            </div>
          </div>
          <div data-v-35dce591="" class="divider" style="display: none;"></div>
        </div>
      </div>
    </div>
  </div>
  <div class="whatsapp-product-list__pagination"><span>1 - 1 de 1</span>
    <div data-v-83fb4bdc="" class="pagination"><button data-v-2d9f2cbf="" data-v-83fb4bdc="" disabled="disabled" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary unnnic-button--icon-on-center">
        <!---->
        <!----><span data-v-ab84ef7c="" data-v-2d9f2cbf="" class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-clean"><svg viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M26.8368 35.0005C26.5868 35.0005 26.3506 34.903 26.1743 34.7255L12.7743 21.3255C12.4206 20.9717 12.2256 20.5017 12.2256 20.0017C12.2256 19.5017 12.4193 19.0305 12.7731 18.6767L26.1743 5.27549C26.3506 5.09799 26.5868 5.00049 26.8368 5.00049C27.0868 5.00049 27.3231 5.09799 27.4993 5.27549C27.6768 5.45174 27.7743 5.68799 27.7743 5.93799C27.7743 6.18799 27.6768 6.42424 27.4993 6.60049L14.0993 20.0005L27.4993 33.4005C27.6768 33.578 27.7743 33.813 27.7743 34.063C27.7743 34.313 27.6768 34.5492 27.4993 34.7255C27.3218 34.9017 27.0868 35.0005 26.8368 35.0005Z" fill="#3B414D" class="primary"></path></svg></span><span data-v-2d9f2cbf="" class="unnnic-button__label">  </span>
        <!---->
      </button><button data-v-2d9f2cbf="" data-v-83fb4bdc="" class="unnnic-button unnnic-button--size-small unnnic-button--secondary">
        <!---->
        <!---->
        <!----><span data-v-2d9f2cbf="" class="unnnic-button__label"> 1 </span>
        <!---->
      </button><button data-v-2d9f2cbf="" data-v-83fb4bdc="" disabled="disabled" class="unnnic-button unnnic-button--size-small unnnic-button--tertiary unnnic-button--icon-on-center">
        <!---->
        <!----><span data-v-ab84ef7c="" data-v-2d9f2cbf="" class="unnnic-icon unnnic-icon__size--sm unnnic-icon-scheme--neutral-clean"><svg viewBox="0 0 40 40" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M13.1635 35.0005C12.9135 35.0005 12.6772 34.903 12.501 34.7255C12.136 34.3605 12.136 33.7655 12.501 33.3992L25.901 20.0005L12.501 6.60049C12.3235 6.42424 12.226 6.18799 12.226 5.93799C12.226 5.68799 12.3235 5.45174 12.501 5.27549C12.6772 5.09799 12.9135 5.00049 13.1635 5.00049C13.4135 5.00049 13.6497 5.09799 13.826 5.27549L27.226 18.6755C27.956 19.4055 27.9572 20.5942 27.2272 21.3242L13.826 34.7255C13.6497 34.903 13.4135 35.0005 13.1635 35.0005Z" fill="#3B414D" class="primary"></path></svg></span><span data-v-2d9f2cbf="" class="unnnic-button__label">  </span>
        <!---->
      </button></div>
  </div>
</div>
`;
